pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -919,6 +919,13 @@ func (p *PriorityQueue) Update(logger klog.Logger, oldPod, newPod *v1.Pod) error
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
+	if p.isSchedulingQueueHintEnabled {
+		if _, ok := p.inFlightPods[newPod.UID]; ok {
```

Comment:
Add a comment that the inflight pod will be requeued using the latest version from the informer cache, which matches what the event delivers.

Reply:
@sanposhiho maybe we need to be smarter?
If the pod update only changed the status, then the pod should probably enter unschedulable or backoff.
But if the pod gets a change in the spec (like a new toleration), the pod should enter active.

Prior to this change, any change would put the pod in active (as it is just inserted into the queue).

Reply:
True. But, I hesitate to just determine whether we put the Pod to activeQ/backoffQ or keep it in the unsched pod pool by whether the update is for .spec or .status. I'm not sure if we have such plugins in in-tree now though, maybe some custom plugins look at Pod's status. Also, https://github.com/kubernetes/enhancements/issues/3990 could be another example of making Pod look at Pod status. (based on the current design)

So, it looks better to determine based on the queueing hint.

```go
// https://github.com/kubernetes/kubernetes/blob/b74e286dc9a458574edb5282524aac70d1da90fc/pkg/scheduler/internal/queue/scheduling_queue.go#L953
	if usPodInfo := p.unschedulablePods.get(newPod); usPodInfo != nil {
		pInfo := updatePod(usPodInfo, newPod)
+		hint := p.isPodWorthRequeuing(logger, pInfo, PodUpdated, usPodInfo, pInfo)
+              // Move this Pod to backoffQ only when QueueingHint suggests.
```

So, let's say NodeA has Taint, Pod is rejected and the Pod goes to the unschedulable Pod pool with TaintToleration plugin as its unschedulable plugin.
So, whenever this Pod is updated, we execute QueueingHint (for PodUpdated event) of TaintToleration. And, when the Pod is updated to get Toleration for that taint, QueueingHint of TaintToleration returns `Queue` and this Pod goes to backoffQ.

Reply:
> determine based on the queueing hint.

If doing so, we would need to make sure all such plugins, whose rejection can be resolved by the update of the Pod being scheduled, have `PodUpdated` event to receive updates of the Pod being scheduled. 
Because, actually, [TaintToleration's EventsToRegister](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go#L59) doesn't have PodUpdated event.

Reply:
Anyway, this improvement could be follow-up in another PR as it'd be too much if we got such stuff in this PR.

Reply:
https://github.com/kubernetes/kubernetes/issues/121183

Reply:
I think we need to do https://github.com/kubernetes/kubernetes/pull/119607/files#r1349613805 in this PR

Otherwise, if a Pod gets a new toleration while it's been scheduled, it will enter the unschedulable queue, and only the flush would get it out of it.

/lgtm cancel

Reply:
> Otherwise, if a Pod gets a new toleration while it's been scheduled, it will enter the unschedulable queue, and only the flush would get it out of it.

Oops, you are right. That's a good point.
@wackxu Here is what you have to do:
- Go through all plugins, and add `PodUpdated` event into EventsToRegister if the change to the Pod could make the Pod schedulable.
  - For example, if the Pod gets a toleration, the Pod may be able to overcome the rejection from TaintToleration plugin. So, you need to add `PodUpdated` event in TaintToleration plugin's EventsToRegister.
- Change [here](https://github.com/kubernetes/kubernetes/blob/b74e286dc9a458574edb5282524aac70d1da90fc/pkg/scheduler/internal/queue/scheduling_queue.go#L953) like [this](https://github.com/kubernetes/kubernetes/pull/119607#discussion_r1349613805).

(The second one could be follow-up in another PR if you want. But, the first one must be included in this PR otherwise the scenario Aldo commented could happen.)

Reply:
cc @kerthcet @Huang-Wei 

We had a discussion of this change in https://github.com/kubernetes/kubernetes/pull/121677#discussion_r1394968673

Summary:
[This idea](https://github.com/kubernetes/kubernetes/pull/119607#discussion_r1349613805) is straightforward, but it has a risk - it could break out-of-tree plugins which rely on the current behavior where all pods are moved out from unschedQ when those Pods are updated. All of those plugins must register PodUpdated event to keep the current behavior. 
We have [an alternative idea](https://github.com/kubernetes/kubernetes/pull/121677#discussion_r1402976336) - we create a new dedicated cluster event (let's name it `PodOwnUpdate` for example). It doesn't break an existing plugin because we put a default QHint if plugins don't have `PodOwnUpdate`. But, it'll be not intuitive, [see this comment](https://github.com/kubernetes/kubernetes/pull/121677#discussion_r1403973443).

So, both solutions are not perfect. We have to accept a breaking change to all plugins (the former idea) or accept a non-straightforward design. (or come up with something new)
@alculquicondor What do you think?

Reply:
I personally prefer a simpler design in kube-scheduler than avoid breaking changes for out-of-tree plugins

Reply:
I'm also that kind of person. Among the current two options, especially to look at this design choice from a long-term PoV, a design that keeps an interface simpler (= use `PodUpdate`) looks like something we should prioritize more than a design that keeps custom plugins safer (= use `PodOwnUpdate`). We can shout it out as a breaking change as we did for the change from `ClusterEvent` to `ClusterEventWithHint`.